plugins/outputs/zabbix/zabbix.go

```diff
@@ -0,0 +1,172 @@
+	SendMetrics(metrics []*zabbix.Metric) (resActive zabbix.Response, resTrapper zabbix.Response, err error)
+	RegisterHost(hostname string, hostMetadata string) error
+}
+
+// Zabbix allows pushing metrics to Zabbix software
+type Zabbix struct {
+	Address                    string          `toml:"address"`
+	AgentActive                bool            `toml:"agent_active"`
+	Prefix                     string          `toml:"prefix"`
+	HostTag                    string          `toml:"host_tag"`
+	SkipMeasurementPrefix      bool            `toml:"skip_measurement_prefix"`
+	LLDSendInterval            config.Duration `toml:"lld_send_interval"`
+	LLDClearInterval           config.Duration `toml:"lld_clear_interval"`
+	Autoregister               string          `toml:"autoregister"`
+	AutoregisterResendInterval config.Duration `toml:"autoregister_resend_interval"`
+	Log                        telegraf.Logger `toml:"-"`
+
+	// lldHandler handles low level discovery data
+	lldHandler zabbixLLD
+	// lldLastSend store the last LLD send to known where to send it again
+	lldLastSend time.Time
+	// autoregisterLastSend stores the last time autoregister data was sent to Zabbix for each host.
+	autoregisterLastSend map[string]time.Time
+	// sender is the interface to send data to Zabbix.
+	sender zabbixSender
+}
+
+//go:embed sample.conf
+var sampleConfig string
+
+func (*Zabbix) SampleConfig() string {
+	return sampleConfig
+}
+
+// Connect does nothing, Write() would initiate connection in each call.
+// Checking if Zabbix server is alive in this step does not allow Telegraf
+// to start if there is a temporal connection problem with the server.
+func (z *Zabbix) Connect() error {
+	return nil
+}
+
+// Init initilizes LLD and autoregister maps. Copy config values to them. Configure Logger.
+func (z *Zabbix) Init() error {
+	// Add port to address if not present
+	if _, _, err := net.SplitHostPort(z.Address); err != nil {
+		z.Address = net.JoinHostPort(z.Address, "10051")
+	}
+
+	z.sender = zabbix.NewSender(z.Address)
+	// Initialize autoregisterLastSend map with size one, as the most common scenario is to have one host.
+	z.autoregisterLastSend = make(map[string]time.Time, 1)
+	z.lldLastSend = time.Now()
+
+	z.lldHandler = zabbixLLD{
+		log:                  z.Log,
+		hostTag:              z.HostTag,
+		clearInterval:        z.LLDClearInterval,
+		lastClear:            time.Now(),
+		cache:                make(map[uint64]bool, 100),
+		receivedData:         make(map[hostLLD]lldTags, 100),
+		previousReceivedData: make(map[hostLLD]uint64, 100),
+	}
+
+	return nil
+}
+
+func (z *Zabbix) Close() error {
+	return nil
+}
+
+// Write sends metrics to Zabbix server
+func (z *Zabbix) Write(metrics []telegraf.Metric) error {
+	if len(metrics) == 0 {
+		return nil
+	}
+
+	zbxMetrics := make([]*zabbix.Metric, 0, len(metrics))
+
+	for _, metric := range metrics {
+		hostname, err := getHostname(z.HostTag, metric)
```

Comment:
What about making `getHostname` part of the `Zabbix` struct?
```suggestion
		hostname, err := z.getHostname(metric)
```

Reply:
``getHostname`` it is also used in ``zabbixLLD.Add`` and in ``TestGetHostname``. Those does not have access to the Zabbix struct.

Reply:
Hmm, but I see you still use a fixed hostname tag in `ZabbixLLD`?

Reply:
Okay, can stay for me.